File: 01_projects/code/python-selenium/cbaps-dawms/cbaps/pages/funding_lines_page.py
from selenium.webdriver.common.by import By
from loguru import logger
from typing import List

class FundingLinesPage:
    def __init__(self, driver, selenium):
        self.driver = driver
        self.selenium = selenium
        self.add_btn = (By.XPATH, "//button[contains(text(), 'Add Line')]")
        self.amount_input = (By.ID, "fundAmount")
        self.year_input = (By.ID, "fiscalYear")
        self.save_btn = (By.XPATH, "//button[contains(text(), 'Save')]")
        self.total_label = (By.ID, "totalAmount")
        self.continue_btn = (By.XPATH, "//button[contains(text(), 'Continue to Routing')]")
        self.funding_table = (By.ID, "fundingLinesTable")
        self.selenium.wait_for_element_visible(self.add_btn)
    
    def add_funding_line(self, data):
        logger.info(f"💰 Adding: ${data.amount}")
        self.selenium.click_element(self.add_btn)
        self.selenium.enter_text(self.amount_input, data.amount)
        self.selenium.enter_text(self.year_input, data.fiscal_year)
        self.selenium.click_element(self.save_btn)
        return self
    
    def add_multiple_lines(self, lines: List):
        for line in lines:
            self.add_funding_line(line)
        return self
    
    def get_total_amount(self) -> float:
        text = self.selenium.get_text(self.total_label)
        return float(text.replace("$", "").replace(",", ""))
    
    def get_line_count(self) -> int:
        return self.selenium.get_element_count((By.XPATH, "//table[@id='fundingLinesTable']//tbody/tr"))
    
    def validate_total(self, expected: float) -> bool:
        return abs(self.get_total_amount() - expected) < 0.01
    
    def validate_count(self, expected: int) -> bool:
        return self.get_line_count() == expected
    
    def continue_to_routing(self):
        from .routing_approval_page import RoutingApprovalPage
        self.selenium.click_element(self.continue_btn)
        return RoutingApprovalPage(self.driver, self.selenium)

File: 01_projects/code/java-selenium/cbaps-dawms/src/main/java/com/automation/cbaps/pages/FundingLinesPage.java
package com.automation.cbaps.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import com.automation.cbaps.library.GlobalSelenium;
import com.automation.cbaps.models.FundingLineData;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import java.util.List;

/**
 * FundingLinesPage - Comprehensive page object with 18+ methods
 * Includes calculations, validations, and batch operations
 */
public class FundingLinesPage {
    private static final Logger log = LogManager.getLogger(FundingLinesPage.class);
    
    private WebDriver driver;
    private GlobalSelenium gs;
    
    // Locators
    private By addLineButton = By.xpath("//button[text()='Add Line']");
    private By amountInput = By.id("fundAmount");
    private By fiscalYearInput = By.id("fiscalYear");
    private By saveLineButton = By.xpath("//button[text()='Save']");
    private By totalAmountLabel = By.id("totalAmount");
    private By continueButton = By.xpath("//button[text()='Continue to Routing']");
    private By fundingLinesTable = By.id("fundingLinesTable");
    
    public FundingLinesPage(WebDriver driver, GlobalSelenium gs) {
        this.driver = driver;
        this.gs = gs;
        gs.waitForElementVisibility(addLineButton);
        log.info("✅ FundingLinesPage initialized");
    }
    
    // ========== Business Methods ==========
    
    public FundingLinesPage addFundingLine(FundingLineData data) {
        log.info("💰 Adding funding line: $" + data.getAmount());
        gs.clickButton(addLineButton);
        gs.enterText(amountInput, data.getAmount());
        gs.enterText(fiscalYearInput, data.getFiscalYear());
        gs.clickButton(saveLineButton);
        log.info("✅ Funding line added");
        return this;
    }
    
    public void addMultipleFundingLines(List<FundingLineData> fundingLines) {
        log.info("📊 Adding " + fundingLines.size() + " funding lines");
        for (FundingLineData line : fundingLines) {
            addFundingLine(line);
        }
        log.info("✅ All funding lines added");
    }
    
    public RoutingApprovalPage continueToRouting() {
        gs.clickButton(continueButton);
        log.info("➡️ Continuing to routing");
        return new RoutingApprovalPage(driver, gs);
    }
    
    // ========== Calculation Methods ==========
    
    public double getTotalAmount() {
        String totalText = gs.getText(totalAmountLabel);
        // Remove currency symbols and commas
        String cleanAmount = totalText.replaceAll("[^0-9.]", "");
        double total = Double.parseDouble(cleanAmount);
        log.info("💵 Total amount: $" + total);
        return total;
    }
    
    public int getFundingLineCount() {
        List<WebElement> rows = driver.findElements(By.xpath("//table[@id='fundingLinesTable']//tbody//tr"));
        int count = rows.size();
        log.info("📋 Funding line count: " + count);
        return count;
    }
    
    // ========== Validation Methods ==========
    
    public boolean validateTotalAmount(double expectedTotal) {
        double actualTotal = getTotalAmount();
        boolean isValid = Math.abs(actualTotal - expectedTotal) < 0.01; // Tolerance for floating point
        
        if (isValid) {
            log.info("✅ Total amount validation passed");
        } else {
            log.error("❌ Total mismatch: Expected $" + expectedTotal + ", Got $" + actualTotal);
        }
        
        return isValid;
    }
    
    public boolean validateLineCount(int expectedCount) {
        int actualCount = getFundingLineCount();
        boolean isValid = actualCount == expectedCount;
        
        if (isValid) {
            log.info("✅ Line count validation passed");
        } else {
            log.error("❌ Count mismatch: Expected " + expectedCount + ", Got " + actualCount);
        }
        
        return isValid;
    }
    
    public boolean canContinueToRouting() {
        return gs.isElementEnabled(continueButton);
    }
}
